src/librustc_errors/emitter.rs

```diff
@@ -268,132 +269,216 @@ pub trait Emitter {
         }
     }
 
-    // This does a small "fix" for multispans by looking to see if it can find any that
-    // point directly at <*macros>. Since these are often difficult to read, this
-    // will change the span to point at the use site.
-    fn fix_multispans_in_std_macros(&self,
-                                    source_map: &Option<Lrc<SourceMap>>,
-                                    span: &mut MultiSpan,
-                                    children: &mut Vec<SubDiagnostic>,
-                                    level: &Level,
-                                    backtrace: bool) {
-        let mut spans_updated = self.fix_multispan_in_std_macros(source_map, span, backtrace);
-        for child in children.iter_mut() {
-            spans_updated |= self.fix_multispan_in_std_macros(
-                                 source_map,
-                                 &mut child.span,
-                                 backtrace
-                             );
+    fn fix_multispans_in_extern_macros_and_render_macro_backtrace(
+        &self,
+        source_map: &Option<Lrc<SourceMap>>,
+        span: &mut MultiSpan,
+        children: &mut Vec<SubDiagnostic>,
+        level: &Level,
+        backtrace: bool,
+    ) {
+        // Check for spans in macros, before `fix_multispans_in_extern_macros`
+        // has a chance to replace them.
+        let has_macro_spans =
+            iter::once(&*span)
+                .chain(children.iter().map(|child| &child.span))
+                .flat_map(|span| span.primary_spans())
+                .copied()
+                .any(|mut sp| {
+                    // FIXME(eddyb) start using `Span::macro_backtrace` whenever
+                    // that becomes cheaper/more useful (see FIXME comment on it),
+                    // instead of having a specialized copy of it here.
```

Comment:
Please, move this logic to a separate function and put it next to `Span::macro_backtrace`.

There are like 4-5 versions of code obtaining the macro backtrace or its specific parts (including `call_span_if_macro` removed in this PR), and I want to unify them, but I won't be able to do it if they are scattered across different crates and inlined.